### ISSUE TITLE
Circuit Breaker does not trigger for queries against tables with big arrays

### ISSUE THREAD

**Issue by jayeff** [crate/crate#11047]
**CrateDB version**: 4.4.1

**Environment description**: CrateDB Cloud

Some statistics on the cluster / table:
3 node cluster running in Bregenz region
table performance_test.arrays with 80k records, 23.2GB
each record has 2 arrays with 20k entries each => ~305kb/record

**Problem description**:

We saw a use case with tables with big arrays (10k-100k entries) and queries against this table can lead to the cluster becoming unresponsive.

If you run queries selecting a larger number of records (5k-10k) these queries either
* throw a RuntimeException[java.lang.OutOfMemoryError: Java heap space]
* or lead to cluster state red with cluster becoming unresponsive for a couple minutes

**Expected behaviour**:

Such queries are caught by a circuit breaker and don't lead to an unresponsive cluster.

**Steps to reproduce**:

Table:

```sql
CREATE TABLE IF NOT EXISTS "performance_test"."arrays" (
   "recorded_date" TIMESTAMP WITH TIME ZONE,
   "values" ARRAY(DOUBLE PRECISION) INDEX OFF,
   "sample_us" ARRAY(BIGINT) INDEX OFF,
   "month" TIMESTAMP WITH TIME ZONE GENERATED ALWAYS AS date_trunc('month', "recorded_date")
)
CLUSTERED INTO 12 SHARDS
PARTITIONED BY ("month")
```

Insert sample data with mkjson + cr8:
(run `cr8 insert-json` multiple times to insert enough data)

```
./mkjson --num 500 values="replicate(20000, randomInt(0, 1000))" sample_us="replicate(20000, randomInt(0, 1000))" recorded_date="randomDateTime('2020-11-01', '2021-01-31 23:00')" > output.json

cr8 insert-json --table performance_test.arrays --hosts localhost:4200 -i output.json
```

Select about 5k-10k records

```
select * from "performance_test"."arrays" LIMIT 5000;
```

There is a preexisting table on presales cluster (`performance_test.arrays`) in Bregenz. You can test the behaviour there.

**Comment by mfussenegger** [comment]
How much memory(+heap) do the nodes have?

**Comment by proddata** [comment]
3x 7 GB ( 3.5 GB heap)
but same behaviour on a
3x 16 GB (4GB heap) 
cluster

**Comment by mfussenegger** [comment]
Thank you for the report. I can reproduce the problem.


Memo to myself:

ReaderBucket gets too big, the fetchSize does take the configured heap into a account but it should probably be more dynamic. We could apply the pattern we use for the `ShardedRequests`

![image](https://user-images.githubusercontent.com/38700/108742996-a1e9bb00-7538-11eb-8a1f-62985869f08a.png)

**Comment by mfussenegger** [comment]
We'll have a fix for this with the next 4.4 release (https://github.com/crate/crate/pull/11054)
